### PR TITLE
FIX: (re-allow) legend OrderedDict handles and labels...

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

As reported in #10262, OrderedDict input no longer works in 2.1.1.  i.e. 

```python
import collections
import matplotlib.pyplot as plt
import numpy as np

X = np.random.randn(10)
Y = np.random.randn(10)
labels = ['a'] * 5 + ['b'] * 5
colors = ['r'] * 5 + ['g'] * 5

fig, ax = plt.subplots()
for x, y, label, color in zip(X, Y, labels, colors):
    ax.scatter(x, y, label=label, c=color)

handles, labels = ax.get_legend_handles_labels()
legend = collections.OrderedDict(zip(labels, handles))
ax.legend(legend.values(), legend.keys(), loc=6, bbox_to_anchor=(1, .5))
```

This PR fixes that (thanks @anntzer) and adds a small test...

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant

### CODE REVIEW

File: lib/matplotlib/legend.py [hunk 1]
```diff
@@ -589,14 +589,17 @@ def __init__(self, parent, handles, labels,
...
+        for label, handle in zip(labels, handles):
+            if (isinstance(label, six.string_types) and
+                    label.startswith('_')):
+                warnings.warn('The handle {!r} has a label of {!r} which '
+                              'cannot be automatically added to the '
+                              'legend.'.format(handle, label))
+            else:
+                _lab.append(label)
+                _hand.append(handle)
+        label, handle = _lab, _hand
```
Comment (anntzer) [inline lib/matplotlib/legend.py]: plural!

Comment (jklymak) [reply lib/matplotlib/legend.py]: Oh duh!

Comment (jklymak) [reply lib/matplotlib/legend.py]: That certainly fixed it alright ;-)

**Comment by tacaswell** [comment]
backported to v2.1.x as c125ab4d2

**Pull request merged**